### PR TITLE
[ADD] web_drop_target

### PR DESCRIPTION
**Pull request by hbrunn**
Drop target support
===================

This module was written to extend the functionality of the web client to support dropping local files into the web client.

By default, this will create an attachment when a file is being dropped on a form, other modules can add more features.

Further, this module is meant as a base module supporting other actions after some file is dropped so that we don't get incompatible drag&drop modules within the OCA.

Usage
=====

To use this module, you need to:

- drag a file from your local computer onto an Odoo form view
- it should become an attachment of the currently opened record

Known issues / Roadmap
======================

* dropping on list or kanban views would be nice too
* handle multiple files
* add an upload progress meter for huge files

### CODE REVIEW

**Comment by NL66278** [comment]
@hbrunn Travis is red due to some phantom.js error

File: web_drop_target/__manifest__.py
Comment (MeetKD) [inline web_drop_target/__manifest__.py:7]: `'website': 'https://github.com/OCA/web',
`

File: web_drop_target/static/src/js/web_drop_target.js
Comment (MeetKD) [inline web_drop_target/static/src/js/web_drop_target.js:97]: Can we add a notification once dropped file is attached and attachment created successfully ?

Comment (hbrunn) [reply web_drop_target/static/src/js/web_drop_target.js:97]: makes a lot of sense. But as this is an extra feature, I'd prefer to wrap this up and have whomever add this who needs it

**Comment by hbrunn** [comment]
@NL66278 now it's only unrelated linting errors which are caused by failing rst syntax error since today

File: web_drop_target/static/src/js/web_drop_target.js
Comment (MeetKD) [reply web_drop_target/static/src/js/web_drop_target.js:97]: :+1:

**Comment by hbrunn** [comment]
I figured I can just as well fix those errors. Now it's green and if both of you approve the PR, we can fast track to have the mani branch green again

**Review by MeetKD — APPROVED** [review]

File: web_drop_target/static/src/js/web_drop_target.js
Comment (simahawk) [reply web_drop_target/static/src/js/web_drop_target.js:97]: you could fire a proper event (via `trigger_up`?), then anybody can hook to it if needed

File: web_drop_target/static/src/js/web_drop_target.js
Comment (simahawk) [inline web_drop_target/static/src/js/web_drop_target.js:86]: can you move attachement values preparation to a separated function? Will make easier to customize.

File: web_widget_x2many_2d_matrix/README.rst
Comment (simahawk) [inline web_widget_x2many_2d_matrix/README.rst:9]: is this here by error?

File: web_drop_target/static/src/js/web_drop_target.js
Comment (hbrunn) [reply web_drop_target/static/src/js/web_drop_target.js:86]: this was what I indended with `extra_data`. Hmmm. A `_prepare_...` function is more canonical I guess

File: web_drop_target/static/src/js/web_drop_target.js
Comment (hbrunn) [reply web_drop_target/static/src/js/web_drop_target.js:97]: triggering an event here is a good thing anyways I think, and then the class itself listens to it in order to `trigger_up` a notification

File: web_widget_x2many_2d_matrix/README.rst
Comment (hbrunn) [reply web_widget_x2many_2d_matrix/README.rst:9]: no, this commit is for making the branch green because yesterday rst syntax warnings were promoted to breaking errors

**Review by tarteo — APPROVED** [review]
https://github.com/OCA/web/tree/10.0/web_chatter_paste can probably use DropTargetMixin to reduce code.

**Review by NL66278 — APPROVED** [review]
:+1:  Code looks good to me as far as I can see. But tested in practice and does what it is supposed to do.

**Review by njeudy — APPROVED** [review]

**Comment by pedrobaeza** [comment]
Has the problems with https://github.com/OCA/social/pull/242 something to do with this PR?

**Comment by njeudy** [comment]
@pedrobaeza I think not, this PR add attachment (as document) not email in chatter .. think it is two separate function ..

**Comment by pedrobaeza** [comment]
This is the foundation for the drag and drop system, so I don't have it clear yet.

**Comment by njeudy** [comment]
@pedrobaeza @hbrunn think web can inspirate with this GPL module :https://github.com/cytex124/odoo-drag-and-drop ?

File: web_drop_target/README.rst
Comment (elicoidal) [inline web_drop_target/README.rst]: +This module extends the functionality of the web client to support dropping local files into the web client.

File: web_drop_target/README.rst
Comment (elicoidal) [inline web_drop_target/README.rst]: By default, an attachment will be created when dropping a file on a form.
Further, this module is meant as a base drag&drop module supporting other actions after some file is dropped so that other modules can add more features.

File: web_drop_target/views/templates.xml
Comment (elicoidal) [inline web_drop_target/views/templates.xml:1]: You can add copyright

**Review by elicoidal — APPROVED** [review]
LGTM

File: web_drop_target/README.rst
Comment (pedrobaeza) [reply web_drop_target/README.rst]: You can take the occasion of using new README fragments system

**Comment by elicoidal** [comment]
It works fine but usability is not very intuitive as is as you do not know when the file is dropped that you can even drop it.
Usually a text or image appears in the drop area confirming that you can drop the file in it.
Other than that excellent feature

**Comment by hbrunn** [comment]
@elicoidal thanks for the review. The addon sets some [css class](https://github.com/OCA/web/pull/907/files#diff-cf5abc94ebf71eacb63a60823a4a986bR18) on the element you drag on, but nothing is defined for this class currently.

I'm definitely much better at technical stuff than at UX, so I just trust someone will fix it with something beautiful once we have merged this

**Comment by MeetKD** [comment]
@pedrobaeza when can we merge this module?

**Comment by pedrobaeza** [comment]
@hbrunn please squash before merging for keeping your GPG signature

**Comment by hbrunn** [comment]
@pedrobaeza done, thanks

File: web_drop_target/static/src/js/web_drop_target.js
Comment (simahawk) [reply web_drop_target/static/src/js/web_drop_target.js:97]: can you drop a comment at least or few lines into ROADMAP to not forget these improvements? :wink:

File: web_drop_target/README.rst
Comment (simahawk) [inline web_drop_target/README.rst:1]: not blocking: png would be better

**Review by simahawk — APPROVED** [review]

File: web_drop_target/README.rst
Comment (hbrunn) [reply web_drop_target/README.rst:1]: the [mtq template](https://raw.githubusercontent.com/OCA/maintainer-quality-tools/master/sample_files/README.md)  disagrees

Comment (simahawk) [reply web_drop_target/README.rst:1]: that's not the one :)
_Was_ here https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst before @sbidoul refactoring. I guess it is handled automatically by the README bot tho.

Comment (pedrobaeza) [reply web_drop_target/README.rst:1]: That's not the template to use. Even more, now we have the README by fragments system: https://github.com/OCA/maintainer-tools/tree/master/template/module/readme, and the badges are auto-generated.

Comment (hbrunn) [reply web_drop_target/README.rst:1]: hmm, can't reply to your other message, so here: Thanks, done in https://github.com/OCA/web/pull/907/files#diff-0bbb40be467186b3f03345f50712b296R33

**Pull request merged**
